<a href="https://colab.research.google.com/github/dhruvi003/ML-Learning/blob/main/flight_price_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

In [3]:
df_train = pd.read_excel('/content/Data_Train.xlsx')
df_train.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [4]:
df_test = pd.read_excel('/content/Test_set.xlsx')
df_test.head()

Airline Date_of_Journey    Source Destination            Route  \
0        Jet Airways       6/06/2019     Delhi      Cochin  DEL → BOM → COK   
1             IndiGo      12/05/2019   Kolkata    Banglore  CCU → MAA → BLR   
2        Jet Airways      21/05/2019     Delhi      Cochin  DEL → BOM → COK   
3  Multiple carriers      21/05/2019     Delhi      Cochin  DEL → BOM → COK   
4           Air Asia      24/06/2019  Banglore       Delhi        BLR → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops              Additional_Info  
0    17:30  04:25 07 Jun  10h 55m      1 stop                      No info  
1    06:20         10:20       4h      1 stop                      No info  
2    19:15  19:00 22 May  23h 45m      1 stop  In-flight meal not included  
3    08:00         21:00      13h      1 stop                      No info  
4    23:55  02:45 25 Jun   2h 50m    non-stop                      No info

In [5]:
df_train.columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price'],
      dtype='object')

In [6]:
df_test.columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info'],
      dtype='object')

In [7]:
df_train.shape

(10683, 11)

In [8]:
df_test.shape

(2671, 10)

In [13]:
final_df = pd.concat([df_train,df_test],axis=1)

In [14]:
final_df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  ...  \
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info  ...   
1    05:50         13:15   7h 25m     2 stops         No info  ...   
2    09:25  04:25 10 Jun      19h     2 stops         No info  ...   
3    18:05         23:30   5h 25m      1 stop         No info  ...   
4    16:50         21:35   4h 45m      1 stop         No info  ...   

             Airline Date_of_Journey    Source Destination            Route  \
0        Jet Airways       6/06/2019     Delhi      Cochin  DEL → BOM → COK   
1             IndiGo      12/05/2019   Kolkata    Banglore  CCU → MAA → BLR   
2        Jet Airways      21/05/2019     Delhi      Cochin  DEL → BOM → COK   
3  Multiple carriers      21/05/2019     Delhi      Cochin  DEL → BOM → COK   
4           Air Asia      24/06/2019  Banglore       Delhi        BLR → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops              Additional_Info  
0    17:30  04:25 07 Jun  10h 55m      1 stop                      No info  
1    06:20         10:20       4h      1 stop                      No info  
2    19:15  19:00 22 May  23h 45m      1 stop  In-flight meal not included  
3    08:00         21:00      13h      1 stop                      No info  
4    23:55  02:45 25 Jun   2h 50m    non-stop                      No info  

[5 rows x 21 columns]

In [15]:
final_df.shape

(10683, 21)

In [19]:
final_df = pd.merge(df_train, df_test, how='outer')
final_df.shape

(13151, 11)

In [21]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13151 entries, 0 to 13150
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13151 non-null  object 
 1   Date_of_Journey  13151 non-null  object 
 2   Source           13151 non-null  object 
 3   Destination      13151 non-null  object 
 4   Route            13150 non-null  object 
 5   Dep_Time         13151 non-null  object 
 6   Arrival_Time     13151 non-null  object 
 7   Duration         13151 non-null  object 
 8   Total_Stops      13150 non-null  object 
 9   Additional_Info  13151 non-null  object 
 10  Price            10697 non-null  float64
dtypes: float64(1), object(10)
memory usage: 1.1+ MB


In [22]:
# from date_of_journey, we need 3 fields
# 01/03/2019
final_df['Date'] = final_df['Date_of_Journey'].str.split('/').str[0]
final_df['Month']= final_df['Date_of_Journey'].str.split('/').str[1]
final_df['Year']= final_df['Date_of_Journey'].str.split('/').str[2]

In [23]:
final_df.columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price', 'Date', 'Month', 'Year'],
      dtype='object')

In [24]:
final_df['Date']=final_df['Date'].astype(int)
final_df['Month']=final_df['Month'].astype(int)
final_df['Year']=final_df['Year'].astype(int)

In [25]:
final_df.drop('Date_of_Journey',axis=1,inplace=True)

In [26]:
final_df.columns

Index(['Airline', 'Source', 'Destination', 'Route', 'Dep_Time', 'Arrival_Time',
       'Duration', 'Total_Stops', 'Additional_Info', 'Price', 'Date', 'Month',
       'Year'],
      dtype='object')

In [28]:
final_df['Arrival_Time'] = final_df['Arrival_Time'].str.split(' ').str[0]

In [29]:
final_df.isnull().sum()

Airline               0
Source                0
Destination           0
Route                 1
Dep_Time              0
Arrival_Time          0
Duration              0
Total_Stops           1
Additional_Info       0
Price              2454
Date                  0
Month                 0
Year                  0
dtype: int64

In [31]:
final_df['Arrivial_hour'] = final_df['Arrival_Time'].str.split(':').str[0].astype(int)
final_df['Arrivial_minutes'] = final_df['Arrival_Time'].str.split(':').str[1].astype(int)

In [32]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13151 entries, 0 to 13150
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Airline           13151 non-null  object 
 1   Source            13151 non-null  object 
 2   Destination       13151 non-null  object 
 3   Route             13150 non-null  object 
 4   Dep_Time          13151 non-null  object 
 5   Arrival_Time      13151 non-null  object 
 6   Duration          13151 non-null  object 
 7   Total_Stops       13150 non-null  object 
 8   Additional_Info   13151 non-null  object 
 9   Price             10697 non-null  float64
 10  Date              13151 non-null  int64  
 11  Month             13151 non-null  int64  
 12  Year              13151 non-null  int64  
 13  Arrivial_hour     13151 non-null  int64  
 14  Arrivial_minutes  13151 non-null  int64  
dtypes: float64(1), int64(5), object(9)
memory usage: 1.5+ MB


In [33]:
final_df.drop('Arrival_Time',axis=1,inplace=True)

In [35]:
final_df['Dept_hour'] = final_df['Dep_Time'].str.split(':').str[0].astype(int)
final_df['Dept_Min'] = final_df['Dep_Time'].str.split(':').str[1].astype(int)

In [37]:
final_df.drop('Dep_Time',axis=1,inplace=True)

In [38]:
# we can drop route and focus on total number of stops
final_df['Total_Stops'].unique()

array(['non-stop', '1 stop', '2 stops', '3 stops', '4 stops', nan],
      dtype=object)

In [39]:
# so we will replace non-stop with 0 etc etc
final_df['Total_Stops']=final_df['Total_Stops'].map({'non-stop':0,'1 stop':1,'2 stops':2,'3 stops': 3, '4 stops': 4})

In [54]:
final_df[final_df['Total_Stops'].isnull()]
# final_df[(final_df['Airline']=='Air India') & (final_df['Source']=='Delhi') & (final_df['Destination']=='Cochin') & (final_df['Price'] > 6500.0) & (final_df['Month']==5) & (final_df['Year']==2019) & ([final_df['Duration'].str.split(' ').str[0]])]['Total_Stops'].value_counts()

Airline Source Destination Route Duration  Total_Stops  \
2128  Air India  Delhi      Cochin   NaN  23h 40m          NaN   

     Additional_Info   Price  Date  Month  Year  Arrivial_hour  \
2128         No info  7480.0     6      5  2019              9   

      Arrivial_minutes  Dept_hour  Dept_Min  
2128                25          9        45

In [62]:
# let's take function to convert duration into minutes
def convert_to_min(time):
    #tims is in 23h 40m
    lst= time.split(' ')
    hr = 0
    min = 0
    if len(lst) == 2:
      hr = lst[0]
      min = lst[1]
      hr =int( hr.split('h')[0])
      min =int( min.split('m')[0])
    elif 'h' in lst[0]:
      hr = lst[0]
      hr =int( hr.split('h')[0])
    else:
      min = lst[0]
      min =int( min.split('m')[0])
    return hr*60 + min

In [63]:
final_df['Duration'].apply(convert_to_min)

0        170
1        165
2        170
3        170
4        165
        ... 
13146    165
13147    155
13148    135
13149    160
13150    170
Name: Duration, Length: 13151, dtype: int64

In [67]:
final_df['Duration'] = final_df['Duration'].apply(convert_to_min)
display(final_df.head())
final_df['Duration']

AttributeError: 'int' object has no attribute 'split'

In [69]:
final_df['Duration'].dtype

dtype('int64')

In [74]:
final_df[final_df['Total_Stops'].isnull()]
final_df[(final_df['Airline']=='Air India') & (final_df['Source']=='Delhi') & (final_df['Destination']=='Cochin') & (final_df['Price'] > 6500.0) & (final_df['Month']==5) & (final_df['Year']==2019) & (final_df['Duration'] > 1200)]['Total_Stops'].value_counts()
final_df['Total_Stops']=final_df['Total_Stops'].fillna(2)

In [75]:
final_df['Total_Stops'].isnull().sum()

np.int64(0)

In [78]:
final_df.columns

Index(['Airline', 'Source', 'Destination', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price', 'Date', 'Month', 'Year', 'Arrivial_hour',
       'Arrivial_minutes', 'Dept_hour', 'Dept_Min'],
      dtype='object')

In [77]:
final_df.drop(['Route','Duration_in_minutes'],axis=1,inplace=True)

In [79]:
final_df['Additional_Info'].unique()
# we will do one hot encoding

array(['No info', 'Red-eye flight', '1 Long layover', '2 Long layover',
       '1 Short layover', 'Change airports', 'No Info', 'Business class',
       'In-flight meal not included', 'No check-in baggage included'],
      dtype=object)

In [81]:
final_df['Airline'].value_counts()
# we will do label-encoding

Airline
Jet Airways                          4677
IndiGo                               2549
Air India                            2135
Multiple carriers                    1493
SpiceJet                             1022
Vistara                               606
Air Asia                              404
GoAir                                 235
Multiple carriers Premium economy      16
Jet Airways Business                    8
Vistara Premium economy                 5
Trujet                                  1
Name: count, dtype: int64

In [82]:
from sklearn.preprocessing import LabelEncoder
labelencoder = LabelEncoder()
final_df['Airline']=labelencoder.fit_transform(final_df['Airline'])
final_df['Source']=labelencoder.fit_transform(final_df['Source'])
final_df['Destination']=labelencoder.fit_transform(final_df['Destination'])
final_df['Additional_Info']=labelencoder.fit_transform(final_df['Additional_Info'])

In [83]:
final_df.shape

(13151, 14)

In [84]:
final_df.head(5)

Airline  Source  Destination  Duration  Total_Stops  Additional_Info  \
0        0       0            5       170          0.0                8   
1        0       0            5       165          0.0                9   
2        0       0            5       170          0.0                8   
3        0       0            5       170          0.0                8   
4        0       0            5       165          0.0                8   

     Price  Date  Month  Year  Arrivial_hour  Arrivial_minutes  Dept_hour  \
0  10715.0     1      3  2019              8                40          5   
1  10873.0     1      3  2019              2                10         23   
2   7074.0     3      3  2019              8                40          5   
3   7652.0     3      3  2019              8                40          5   
4   6135.0     3      3  2019              2                10         23   

   Dept_Min  
0        50  
1        25  
2        50  
3        50  
4        25

In [87]:
df_train = final_df[~final_df['Price'].isnull()]
df_test = final_df[final_df['Price'].isnull()]

# Task
Build and train a model to predict flight prices using the `final_df` dataframe.

## Feature selection

### Subtask:
Define the features (X) and the target variable (y) for the model.


**Reasoning**:
Define the features (X) and the target variable (y) for the model using the df_train dataframe.



In [88]:
X = df_train.drop('Price', axis=1)
y = df_train['Price']

## Model selection

### Subtask:
Choose a suitable regression model for predicting flight prices.


**Reasoning**:
Import necessary regression models from scikit-learn.



In [89]:
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor

## Model training

### Subtask:
Train the selected regression models using the training data.


**Reasoning**:
Instantiate and fit the selected regression models to the training data.



In [90]:
lr = LinearRegression()
rf = RandomForestRegressor(random_state=42)
dt = DecisionTreeRegressor(random_state=42)

lr.fit(X, y)
rf.fit(X, y)
dt.fit(X, y)

DecisionTreeRegressor(random_state=42)

## Model evaluation

### Subtask:
Evaluate the performance of the trained regression models using appropriate metrics.


**Reasoning**:
Import the necessary evaluation metrics and then evaluate the performance of the trained models on the training data by calculating and printing the MSE and R2 scores for each model.



In [100]:
df_test.drop(['predicted_price_dt','predicted_price_lr','predicted_price_rf'],axis=1,inplace=True)

## Prediction on test data

### Subtask:
Use the trained models to predict prices on the test dataset (`df_test`).


**Reasoning**:
Use the trained models to predict prices on the test dataset and store the predictions in new columns.



In [102]:
# Reload df_test to remove any previously added columns
df_test = final_df[final_df['Price'].isnull()].copy()

# Ensure df_test has the same columns as X (features used for training)
X_test = df_test.drop('Price', axis=1).copy()

# Make predictions with Linear Regression
df_test['predicted_price_lr'] = lr.predict(X_test)

# Make predictions with Random Forest Regressor
df_test['predicted_price_rf'] = rf.predict(X_test)

# Make predictions with Decision Tree Regressor
df_test['predicted_price_dt'] = dt.predict(X_test)

display(df_test.head())

Airline  Source  Destination  Duration  Total_Stops  Additional_Info  \
12        0       3            0       150          0.0                8   
14        0       0            2       165          0.0                8   
15        0       0            2       170          0.0                8   
17        0       3            0       155          0.0                8   
21        0       3            0       955          1.0                8   

    Price  Date  Month  Year  Arrivial_hour  Arrivial_minutes  Dept_hour  \
12    NaN     1      3  2019              1                20         22   
14    NaN     1      4  2019             13                55         11   
15    NaN     1      4  2019              2                45         23   
17    NaN     1      4  2019             12                55         10   
21    NaN     1      4  2019             23                30          7   

    Dept_Min  predicted_price_lr  predicted_price_rf  predicted_price_dt  
12        50         6339.245362             6004.60              5515.0  
14        10         6393.219197             4169.74              4483.0  
15        55         6723.295762             4406.59              4483.0  
17        20         5206.391703             4201.00              4107.0  
21        35         9870.627451             5705.08              5192.0

## Summary:

### Data Analysis Key Findings

*   Linear Regression had the highest Mean Squared Error (MSE) of 12,020,612.90 and the lowest R-squared of 0.43 on the training data, indicating a poor fit.
*   Random Forest Regressor achieved a much lower MSE of 442,320.54 and a high R-squared of 0.98 on the training data.
*   Decision Tree Regressor showed the lowest MSE of 103,737.93 and a perfect R-squared of 1.00 on the training data, which might suggest overfitting.
*   Successful prediction on the test data required ensuring the feature columns exactly matched those used during training.

### Insights or Next Steps

*   Evaluate the models on a separate test set to get a realistic measure of their performance and check for overfitting, especially with the Decision Tree Regressor.
*   Consider hyperparameter tuning for the Random Forest Regressor and Decision Tree Regressor to potentially improve performance and mitigate overfitting.


In [103]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

# Split the training data into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# Retrain the models on the training split (using X_train and y_train)
lr.fit(X_train, y_train)
rf.fit(X_train, y_train)
dt.fit(X_train, y_train)

# Evaluate Linear Regression on the validation set (using X_val and y_val)
y_pred_lr = lr.predict(X_val)
mse_lr = mean_squared_error(y_val, y_pred_lr)
r2_lr = r2_score(y_val, y_pred_lr)
print(f"Linear Regression - MSE: {mse_lr:.2f}, R2: {r2_lr:.2f}")

# Evaluate Random Forest Regressor on the validation set (using X_val and y_val)
y_pred_rf = rf.predict(X_val)
mse_rf = mean_squared_error(y_val, y_pred_rf)
r2_rf = r2_score(y_val, y_pred_rf)
print(f"Random Forest Regressor - MSE: {mse_rf:.2f}, R2: {r2_rf:.2f}")

# Evaluate Decision Tree Regressor on the validation set (using X_val and y_val)
y_pred_dt = dt.predict(X_val)
mse_dt = mean_squared_error(y_val, y_pred_dt)
r2_dt = r2_score(y_val, y_pred_dt)
print(f"Decision Tree Regressor - MSE: {mse_dt:.2f}, R2: {r2_dt:.2f}")

Linear Regression - MSE: 16991047.33, R2: 0.36
Random Forest Regressor - MSE: 6411419.46, R2: 0.76
Decision Tree Regressor - MSE: 7733440.67, R2: 0.71


In [105]:
# Drop the previously added prediction columns from df_test if they exist
df_test = df_test.drop(['predicted_price_lr', 'predicted_price_rf', 'predicted_price_dt'], axis=1, errors='ignore')

# Ensure df_test has the same columns as X (features used for training)
X_test = df_test.drop('Price', axis=1).copy()

# Make predictions with Linear Regression
df_test['predicted_price_lr'] = lr.predict(X_test)

# Make predictions with Random Forest Regressor
df_test['predicted_price_rf'] = rf.predict(X_test)

# Make predictions with Decision Tree Regressor
df_test['predicted_price_dt'] = dt.predict(X_test)

display(df_test.head())

Airline  Source  Destination  Duration  Total_Stops  Additional_Info  \
12        0       3            0       150          0.0                8   
14        0       0            2       165          0.0                8   
15        0       0            2       170          0.0                8   
17        0       3            0       155          0.0                8   
21        0       3            0       955          1.0                8   

    Price  Date  Month  Year  Arrivial_hour  Arrivial_minutes  Dept_hour  \
12    NaN     1      3  2019              1                20         22   
14    NaN     1      4  2019             13                55         11   
15    NaN     1      4  2019              2                45         23   
17    NaN     1      4  2019             12                55         10   
21    NaN     1      4  2019             23                30          7   

    Dept_Min  predicted_price_lr  predicted_price_rf  predicted_price_dt  
12        50         6413.446351             6202.75              5515.0  
14        10         6350.584829             4401.84              4483.0  
15        55         6685.458000             4919.10              4483.0  
17        20         5284.294768             4190.16              4107.0  
21        35         9944.221158             5385.97              5192.0

## Summary:

### Data Analysis Key Findings

*   The models were evaluated on a validation set (20% of the original training data) to get a more realistic measure of performance and check for overfitting.
*   Linear Regression had an MSE of 16991047.33 and an R2 of 0.36 on the validation set, indicating it did not perform well.
*   Random Forest Regressor achieved a significantly lower MSE of 6411419.46 and a higher R2 of 0.76 on the validation set, showing better performance than Linear Regression.
*   Decision Tree Regressor had an MSE of 7733440.67 and an R2 of 0.71 on the validation set. While its performance on the training data showed perfect R2 (potentially indicating overfitting), the validation set results give a more realistic picture.

### Insights or Next Steps

*   Based on the validation set evaluation, the Random Forest Regressor appears to be the best performing model among the three we trained.
*   Further hyperparameter tuning of the Random Forest Regressor could potentially improve its performance on unseen data.
*   The predicted prices for the test dataset (`df_test`) are now available in the columns `predicted_price_lr`, `predicted_price_rf`, and `predicted_price_dt`. You can choose to use the predictions from the best performing model (Random Forest Regressor) for your task.

# Task
Train and evaluate Gradient Boosting Regressor and Support Vector Regressor models on the provided training and validation data, predict prices on the test dataset ("df_test"), and update the model performance summary.

## Model selection

### Subtask:
Add Gradient Boosting Regressor and Support Vector Regressor to the list of models.


**Reasoning**:
Import Gradient Boosting Regressor and Support Vector Regressor for model training.



In [106]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.svm import SVR

## Model training

### Subtask:
Train the new regression models using the training data (`X_train`, `y_train`).


**Reasoning**:
Instantiate and fit the Gradient Boosting Regressor and SVR models to the training data.



In [107]:
gbr = GradientBoostingRegressor(random_state=42)
svr = SVR()

gbr.fit(X_train, y_train)
svr.fit(X_train, y_train)

SVR()

## Model evaluation

### Subtask:
Evaluate the performance of the new trained models on the validation set (`X_val`, `y_val`) using appropriate metrics (MSE and R2).


**Reasoning**:
Evaluate the performance of the newly trained models on the validation set using MSE and R2 scores.



In [108]:
y_pred_gbr = gbr.predict(X_val)
mse_gbr = mean_squared_error(y_val, y_pred_gbr)
r2_gbr = r2_score(y_val, y_pred_gbr)
print(f"Gradient Boosting Regressor - MSE: {mse_gbr:.2f}, R2: {r2_gbr:.2f}")

y_pred_svr = svr.predict(X_val)
mse_svr = mean_squared_error(y_val, y_pred_svr)
r2_svr = r2_score(y_val, y_pred_svr)
print(f"Support Vector Regressor - MSE: {mse_svr:.2f}, R2: {r2_svr:.2f}")

Gradient Boosting Regressor - MSE: 6041970.75, R2: 0.77
Support Vector Regressor - MSE: 25875960.54, R2: 0.03


## Prediction on test data

### Subtask:
Use the new trained models to predict prices on the test dataset (`df_test`).


**Reasoning**:
Use the trained Gradient Boosting Regressor and Support Vector Regressor models to predict prices on the test dataset and store the predictions in new columns in df_test.



In [109]:
# Make predictions with Gradient Boosting Regressor
df_test['predicted_price_gbr'] = gbr.predict(X_test)

# Make predictions with Support Vector Regressor
df_test['predicted_price_svr'] = svr.predict(X_test)

display(df_test.head())

Airline  Source  Destination  Duration  Total_Stops  Additional_Info  \
12        0       3            0       150          0.0                8   
14        0       0            2       165          0.0                8   
15        0       0            2       170          0.0                8   
17        0       3            0       155          0.0                8   
21        0       3            0       955          1.0                8   

    Price  Date  Month  Year  Arrivial_hour  Arrivial_minutes  Dept_hour  \
12    NaN     1      3  2019              1                20         22   
14    NaN     1      4  2019             13                55         11   
15    NaN     1      4  2019              2                45         23   
17    NaN     1      4  2019             12                55         10   
21    NaN     1      4  2019             23                30          7   

    Dept_Min  predicted_price_lr  predicted_price_rf  predicted_price_dt  \
12        50         6413.446351             6202.75              5515.0   
14        10         6350.584829             4401.84              4483.0   
15        55         6685.458000             4919.10              4483.0   
17        20         5284.294768             4190.16              4107.0   
21        35         9944.221158             5385.97              5192.0   

    predicted_price_gbr  predicted_price_svr  
12          9990.472758          7841.740660  
14          3910.050421          7853.674790  
15          4216.506745          7857.669953  
17          3741.775741          7845.601560  
21          6037.209431          8591.232467